Build fails on last stage

The problem seems to be that `make sdist` needs to create `__git__.py` file in order to include it in the distribution, however `.git` folder is not present by that point. We could either perform `sdist` generation some time earlier when `.git` was still present, or just make sure that `.git` is properly stashed with the rest of the distro.
```
[build-sdist] Running shell script
+ make CI=1 ubuntu_build_sdist_in_docker
docker run \
        --rm \
        --init \
        -u `id -u`:`id -g` \
        -v `pwd`:/dot \
        -w /dot \
        --entrypoint /bin/bash \
        -e "CI_VERSION_SUFFIX=master_232" \
        -e "DT_LARGE_TESTS_ROOT=" \
         \
        docker.h2o.ai/opsh2oai/datatable-build-x86_64_ubuntu:0.6.0-PR-1210.1 \
        -c ". /envs/datatable-py36-with-pandas/bin/activate && \
            python --version && \
            make CI=1 sdist"
Unable to find image 'docker.h2o.ai/opsh2oai/datatable-build-x86_64_ubuntu:0.6.0-PR-1210.1' locally
0.6.0-PR-1210.1: Pulling from opsh2oai/datatable-build-x86_64_ubuntu
Digest: sha256:e0f695d248a4fcc51646f9338ad1f257a4e2ceed1d79ac98c360494979f3b2af
Status: Downloaded newer image for docker.h2o.ai/opsh2oai/datatable-build-x86_64_ubuntu:0.6.0-PR-1210.1
Python 3.6.5
fatal: Not a git repository (or any parent up to mount point /dot)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
python setup.py sdist
Setting environment variables:
  CC = /clang+llvm-6.0.1-x86_64-linux-gnu-ubuntu-16.04/bin/clang++ -isystem /usr/include/python3.6m
  CXX = /clang+llvm-6.0.1-x86_64-linux-gnu-ubuntu-16.04/bin/clang++ -isystem /usr/include/python3.6m
  LDFLAGS = 
  ARCHFLAGS = -m64
  LLVM_CONFIG = /clang+llvm-6.0.1-x86_64-linux-gnu-ubuntu-16.04/bin/llvm-config
Cannot determine git revision of the package because folder .git is missing. Current directory: /dot
Makefile:126: recipe for target 'sdist' failed
make: *** [sdist] Error 1
make: *** [ubuntu_build_sdist_in_docker] Error 2
script returned exit code 2
```

--- a/Makefile
+++ b/Makefile
@@ -195,6 +195,7 @@ centos7_in_docker: Dockerfile-centos7.$(PLATFORM).tag
 		-w /dot \
 		--entrypoint /bin/bash \
 		-e "CI_VERSION_SUFFIX=$(CI_VERSION_SUFFIX)" \
+		-e "DTBL_GIT_HASH=$(DTBL_GIT_HASH)" \
 		$(CONTAINER_NAME_TAG) \
 		-c 'make dist'
 	mkdir -p $(DIST_DIR)/$(PLATFORM)
@@ -233,6 +234,7 @@ centos7_build_in_docker_impl:
 		-w /dot \
 		--entrypoint /bin/bash \
 		-e "CI_VERSION_SUFFIX=$(CI_VERSION_SUFFIX)" \
+		-e "DTBL_GIT_HASH=$(DTBL_GIT_HASH)" \
 		$(CUSTOM_ARGS) \
 		$(CENTOS_DOCKER_IMAGE_NAME) \
 		-c ". activate $(BUILD_VENV) && \
@@ -253,6 +255,7 @@ centos7_version_in_docker:
 		-w /dot \
 		--entrypoint /bin/bash \
 		-e "CI_VERSION_SUFFIX=$(CI_VERSION_SUFFIX)" \
+		-e "DTBL_GIT_HASH=$(DTBL_GIT_HASH)" \
 		$(CUSTOM_ARGS) \
 		$(CENTOS_DOCKER_IMAGE_NAME) \
 		-c ". activate datatable-py36-with-pandas && \
@@ -270,6 +273,7 @@ centos7_test_in_docker_impl:
 		-w /dot \
 		--entrypoint /bin/bash \
 		-e "DT_LARGE_TESTS_ROOT=$(DT_LARGE_TESTS_ROOT)" \
+		-e "DTBL_GIT_HASH=$(DTBL_GIT_HASH)" \
 		$(CUSTOM_ARGS) \
 		$(CENTOS_DOCKER_IMAGE_NAME) \
 		-c ". activate $(TEST_VENV) && \
@@ -300,6 +304,7 @@ ubuntu_build_in_docker_impl:
 		--entrypoint /bin/bash \
 		-e "CI_VERSION_SUFFIX=$(CI_VERSION_SUFFIX)" \
 		-e "DT_LARGE_TESTS_ROOT=$(DT_LARGE_TESTS_ROOT)" \
+		-e "DTBL_GIT_HASH=$(DTBL_GIT_HASH)" \
 		$(CUSTOM_ARGS) \
 		$(UBUNTU_DOCKER_IMAGE_NAME) \
 		-c ". /envs/$(BUILD_VENV)/bin/activate && \
@@ -316,6 +321,7 @@ ubuntu_build_sdist_in_docker:
 		--entrypoint /bin/bash \
 		-e "CI_VERSION_SUFFIX=$(CI_VERSION_SUFFIX)" \
 		-e "DT_LARGE_TESTS_ROOT=$(DT_LARGE_TESTS_ROOT)" \
+		-e "DTBL_GIT_HASH=$(DTBL_GIT_HASH)" \
 		$(CUSTOM_ARGS) \
 		$(UBUNTU_DOCKER_IMAGE_NAME) \
 		-c ". /envs/datatable-py36-with-pandas/bin/activate && \
@@ -337,6 +343,7 @@ ubuntu_coverage_py36_with_pandas_in_docker:
 		-w /dot \
 		--entrypoint /bin/bash \
 		-e "DT_LARGE_TESTS_ROOT=$(DT_LARGE_TESTS_ROOT)" \
+		-e "DTBL_GIT_HASH=$(DTBL_GIT_HASH)" \
 		$(CUSTOM_ARGS) \
 		$(UBUNTU_DOCKER_IMAGE_NAME) \
 		-c ". /envs/datatable-py36-with-pandas/bin/activate && \
@@ -352,6 +359,7 @@ ubuntu_test_in_docker_impl:
 		-w /dot \
 		--entrypoint /bin/bash \
 		-e "DT_LARGE_TESTS_ROOT=$(DT_LARGE_TESTS_ROOT)" \
+		-e "DTBL_GIT_HASH=$(DTBL_GIT_HASH)" \
 		$(CUSTOM_ARGS) \
 		$(UBUNTU_DOCKER_IMAGE_NAME) \
 		-c ". /envs/$(TEST_VENV)/bin/activate && \

--- a/ci/Jenkinsfile2
+++ b/ci/Jenkinsfile2
@@ -485,26 +485,22 @@ ansiColor('xterm') {
                 parallel(testStages)
             }
             // Build sdist
-
-            // FIXME
-            // if (doPublish()) {
-                node(X86_64_BUILD_NODE_LABEL) {
-                    def stageDir = 'build-sdist'
-                    buildSummary.stageWithSummary ('Build sdist', stageDir) {
-                        cleanWs()
-                        dumpInfo()
-                        withEnv(["CI_VERSION_SUFFIX=${CI_VERSION_SUFFIX}", "DTBL_GIT_HASH=${gitHash}"]) {
-                            dir (stageDir) {
-                                unstash 'datatable-sources'
-                                unstash 'VERSION'
-                                sh "make ${MAKE_OPTS} ubuntu_build_sdist_in_docker"
-                                stash includes: 'dist/*.tar.gz', name: 'sdist-tar'
-                                arch "dist/*.tar.gz"
-                            }
+            node(X86_64_BUILD_NODE_LABEL) {
+                def stageDir = 'build-sdist'
+                buildSummary.stageWithSummary ('Build sdist', stageDir) {
+                    cleanWs()
+                    dumpInfo()
+                    withEnv(["CI_VERSION_SUFFIX=${CI_VERSION_SUFFIX}", "DTBL_GIT_HASH=${gitHash}"]) {
+                        dir (stageDir) {
+                            unstash 'datatable-sources'
+                            unstash 'VERSION'
+                            sh "make ${MAKE_OPTS} ubuntu_build_sdist_in_docker"
+                            stash includes: 'dist/*.tar.gz', name: 'sdist-tar'
+                            arch "dist/*.tar.gz"
                         }
                     }
                 }
-            // }
+            }
             // Publish snapshot to S3
             if (doPublish()) {
                 node(RELEASE_NODE_LABEL) {

--- a/ci/Jenkinsfile2
+++ b/ci/Jenkinsfile2
@@ -69,6 +69,8 @@ def needsLargerTest
 def dockerArgs = createDockerArgs()
 // String with current version
 def versionText
+// String with current git revision
+def gitHash
 
 MAKE_OPTS = "CI=1"
 
@@ -98,7 +100,8 @@ ansiColor('xterm') {
                 dir (stageDir) {
                     buildSummary.stageWithSummary('Checkout', stageDir) {
                         def scmEnv = checkout scm
-                        env.BRANCH_NAME = scmEnv['GIT_BRANCH'].replaceAll('origin/', '').replaceAll('/', '_')
+                        gitHash = scmEnv.GIT_COMMIT
+                        env.BRANCH_NAME = scmEnv.GIT_BRANCH.replaceAll('origin/', '').replaceAll('/', '_')
                         if (doPPC()) {
                             manager.addBadge("success.gif", "PPC64LE build triggered!")
                         }
@@ -482,13 +485,15 @@ ansiColor('xterm') {
                 parallel(testStages)
             }
             // Build sdist
-            if (doPublish()) {
+
+            // FIXME
+            // if (doPublish()) {
                 node(X86_64_BUILD_NODE_LABEL) {
                     def stageDir = 'build-sdist'
                     buildSummary.stageWithSummary ('Build sdist', stageDir) {
                         cleanWs()
                         dumpInfo()
-                        withEnv(["CI_VERSION_SUFFIX=${CI_VERSION_SUFFIX}"]) {
+                        withEnv(["CI_VERSION_SUFFIX=${CI_VERSION_SUFFIX}", "DTBL_GIT_HASH=${gitHash}"]) {
                             dir (stageDir) {
                                 unstash 'datatable-sources'
                                 unstash 'VERSION'
@@ -499,7 +504,7 @@ ansiColor('xterm') {
                         }
                     }
                 }
-            }
+            // }
             // Publish snapshot to S3
             if (doPublish()) {
                 node(RELEASE_NODE_LABEL) {

--- a/setup.py
+++ b/setup.py
@@ -84,16 +84,21 @@ def get_test_dependencies():
 
 def make_git_version_file(force):
     import subprocess
-    if not os.path.isdir(".git"):
-        if force:
-            raise SystemExit("Cannot determine git revision of the package "
-                             "because folder .git is missing. Current "
-                             "directory: %s"
-                             % os.path.abspath(os.path.curdir))
-        print("Could not generate __git__.py", file=stderr)
-        return
-    out = subprocess.check_output(["git", "rev-parse", "HEAD"])
-    githash = out.decode("ascii").strip()
+    # Try to read git revision from env.
+    githash = os.getenv('DTBL_GIT_HASH', None)
+    if not githash:
+        # Need to get git revision from git cmd. Fail if .git dir is not accessible.
+        if not os.path.isdir(".git"):
+            if force:
+                raise SystemExit("Cannot determine git revision of the package "
+                                 "because folder .git is missing and DTBL_GIT_HASH "
+                                 "is not set. Current directory: %s"
+                                 % os.path.abspath(os.path.curdir))
+            print("Could not generate __git__.py", file=stderr)
+            return
+        # Read git revision using cmd.
+        out = subprocess.check_output(["git", "rev-parse", "HEAD"])
+        githash = out.decode("ascii").strip()
     print("Generating __git__.py with hash=%s\n" % githash, file=stderr)
     with open("datatable/__git__.py", "w", encoding="utf-8") as o:
         o.write(

